### PR TITLE
Adding support for filesystem size

### PR DESCRIPTION
**Pull request by mrbean-bremen**
- added parameter to set the total size of the filesystem
- keep track of used size for added and deleted files
- raise IOError for insufficient space on creating or resizing a file
- added shutil.disk_usage() implementation
- see issue #86

### CODE REVIEW

File: fake_filesystem_test.py [hunk 1]
```diff
@@ -3250,5 +3250,65 @@ def testExistsWithMixedSeparators(self):
     self.assertTrue(self.filesystem.Exists('?foo?bar?xyzzy?plugh'))
 
 
+class DiskSpaceTest(TestCase):
+  def setUp(self):
+    self.filesystem = fake_filesystem.FakeFilesystem(path_separator='/', total_size=1024)
+
+  def testFileSystemSizeAfterFileCreation(self):
```
Comment (jmcgeheeiv) [inline fake_filesystem_test.py]: File contents are commonly strings.

Please add a test trying the ASCII string 'xyzzy', I suppose that this should consume:

> 5 characters + null terminator = 6 bytes

Then add a test trying the Unicode string '複雑' which means "complicated" in Japanese and Chinese.  I would think this would consume:

> 3 bytes for 複 + 3 bytes for 雑 + null terminator = 7 bytes

File: pyfakefs/fake_filesystem.py [hunk 2]
```diff
@@ -248,38 +255,76 @@ def SetLargeFileSize(self, st_size):
...
       contents = Hexlified(contents)
 
-    self.st_ctime = time.time()
-    self.st_mtime = self._st_ctime
-    self.st_size = len(contents)
+    if self.contents:
+      self.SetSize(0)
+    current_size = self.st_size or 0
     self.contents = contents
+    self.st_size = len(self.contents)
```
Comment (jmcgeheeiv) [inline pyfakefs/fake_filesystem.py]: This gives the wrong answer for non-ascii strings:

    >>> len('複雑')
    2

Maybe you need something like this:

    >>> len('複雑'.encode("utf8"))
    6

I tried actually writing this to a file, and Linux `ls -l` says the file size is 6.

**Comment by mrbean-bremen** [comment]
Are you sure about the null terminator? AFAIK, this is not added automatically if writing python strings. About the unicode strings: you are right, I will add a test, and fix the code, currently it is just taking len(contents), which will not work correctly with unicode strings. 
But I had thought that unicode strings in both Python 2 and 3 are written as UCS-2 (or UCS-4, but your nice Kanji shall fit into the 16 bit code point range), which would mean 2 bytes per character. Anyway, I will check this...

**Comment by mrbean-bremen** [comment]
Ok, didn't see your second post as I answered the first one. The problem I see here is that you encoded the string in UTF-8 before writing it to the disk, which is not a given, as I understand it. But I think I have to better understand the code first - especially how reading and writing is done in the fake file system. I will ask you if I need some help there... thanks for your support.

**Comment by mrbean-bremen** [comment]
Ok, I made a few tests - you are right about UTF-8, I didn't know that. Under Python 2, strings are indeed written as UTF-8. Under Python 3, if the file is opened binary, you have to provide a bytestring (e.g. do the encoding yourself, as you did), otherwise it depends on your text file encoding. I will have a go at this somewhere in the next days.

**Comment by jmcgeheeiv** [comment]
I read/write Japanese and I am relatively familiar with non-ASCII, so if you need anything I am at your service.

Indeed I was wrong about the null terminator.  I removed my comments suggesting it might be necessary.

**Comment by mrbean-bremen** [comment]
Ok, I hopefully fixed this - please check if I shall add more tests. And if I ever want to learn Japanese, I will get back to you :)

File: fake_filesystem_test.py [hunk 3]
```diff
@@ -3276,5 +3276,84 @@ def testSplitPath(self):
...
+
+  def testFileSystemSizeAfterBinaryFileCreation(self):
+    self.filesystem.CreateFile('/foo/bar', contents=b'xyzzy')
+    self.assertEqual((100, 5, 95), self.filesystem.GetDiskUsage())
+
+  def testFileSystemSizeAfterAsciiStringFileCreation(self):
+    self.filesystem.CreateFile('/foo/bar', contents='complicated')
+    self.assertEqual((100, 11, 89), self.filesystem.GetDiskUsage())
+
+  def testFileSystemSizeAfter2ByteUnicodeStringFileCreation(self):
```
Comment (jmcgeheeiv) [inline fake_filesystem_test.py]: Two byte UTF-8, very cool.
